students/laura_denney/lesson09/mailroom_objectoriented.py

```diff
@@ -0,0 +1,280 @@
+                (other.sum_donations, other.name))
+
+    def append(self, donation):
+        self.donations.append(donation)
+####################################################
+#Donors Class - manipulates a list of donors
+class Donors(object):
+    donor_list = None
+
+    def __init__(self, *args):
+        self.donor_list = []
+        for arg in args:
+            self.donor_list.append(arg)
+
+    def __repr__(self):
+        repr_str = "Donors(" + ", ".join(["{}"] * len(self)) + ")"
+        return repr_str.format(*self.donor_list)
+
+    def __len__(self):
+        return len(self.donor_list)
+
+    def add_new_donor(self, donor):
+        self.donor_list.append(donor)
+
+
+    def is_current_donor(self, donor):
+        return donor.lower() in repr(self)
+
+    def get_list(self):
+        names = [donor.name.title() for donor in self.donor_list]
+        strformat = "\nWe have {} current donors: " + ", ".join(["{}"] * len(names))
+        return strformat.format(len(names), *names)
+
+    def create_string_report(self):
+        self.donor_list.sort(reverse = True)
+        report = '''
+Donor Name                | Total Given | Num Gifts | Average Gift
+------------------------------------------------------------------'''
+        strformat = '\n{:<26}${:>13.2f}{:^12}${:>13.2f}'
+        for donor in self.donor_list:
+            report +=  strformat.format(donor.name.title(),donor.sum_donations,
+                                    donor.total_donations, donor.average_donation)
+        return report
+
+    def get_donor(self, name):
+        for donor in self.donor_list:
+            if donor.name == name:
+                return donor
+
+    def check_current_donor(self, name, donation):
+        if self.is_current_donor(name):
+            current_donor = self.get_donor(name)
+            current_donor.append(donation)
+            return "\n{} is a current donor, we will update their donations.".format(name.title())
+
+        else:
+            self.add_new_donor(Donor(name, donation))
+            return "\n{} is not a current donor, we will add them to our system.".format(name.title())
+
+    def send_letters_per_donor(self):
+        total_letters = 0
+        for each_donor in self.donor_list:
+            donor = each_donor.name.title()
+            #Name formatting for one word names like 'Cher' vs normal full names
+            first_last = donor.split(" ")
+            if len(first_last) == 1:
+                first = first_last[0]
+                last = ""
+            else:
+                first = first_last[0]
+                last = first_last[1]
+            total_letters += send_email(donor, each_donor.sum_donations, "{}_{}_{}.txt".format(first, last, date.today().isoformat()))
+        return total_letters
+
+    @classmethod
+    def load_list_donors(cls):
+        with open("donor_list.txt", "r") as infile:
+            infile_str = infile.read()
+        self = eval(infile_str)
```

Comment:
I would avoid using `eval`

Use csv format or JSON String - but all of the various ways to save Donors will be covered in the next course 🙂